src/shadowbox/infrastructure/logging.ts

```diff
@@ -51,18 +51,60 @@ function makeLogMessage(level: LevelPrefix, callsite: Callsite, message: string)
   )}:${callsite.getLineNumber()}] ${message}`;
 }
 
+export enum LogLevel {
+  // The order here is important, from less to more verbose.
+  ERROR,
+  WARNING,
+  INFO,
+  DEBUG,
+}
+
+const maxMsgLevel = logLevelFromEnvironment();
+
+function logLevelFromEnvironment(): LogLevel {
+  if (process.env.LOG_LEVEL) {
+    return parseLogLevel(process.env.LOG_LEVEL);
+  } else {
+    return LogLevel.INFO;
+  }
+}
+
+function parseLogLevel(levelStr: string) {
+  switch(levelStr.toLowerCase()) {
+    case "error":
+      return LogLevel.ERROR;
+    case "warning":
+    case "warn":
+      return LogLevel.WARNING;
+    case "info":
+      return LogLevel.INFO;
+    case "debug":
+      return LogLevel.DEBUG;
+    default:
+      throw new Error(`Invalid log level "${levelStr}"`);
+  }
+}
+
 export function error(message: string) {
-  console.error(makeLogMessage('E', getCallsite(), message));
+  if (LogLevel.ERROR <= maxMsgLevel) {
```

Comment:
I find it more readable to see the condition for which the code applies. I'd agree with you if there were multiple conditions to escape, to avoid nesting.